### PR TITLE
feat: Added client anticipation sample [MTT-8052]

### PR DESCRIPTION
**Pull request by ShadauxCat**
### Description

This adds a sample to show off client anticipation. It shows a few things:
- A network variable that can be anticipated and will react to an incorrect anticipation by snapping to the new value
- A network variable that can be anticipated and will react to an incorrect anticipation by interpolating to the new value
- A network variable that is controlled by the server, where the client constantly anticipates a new value to hide latency by doing its own extra simulation based on the current latency
- Anticipation of player input by having a player object whose transform is fully server authoritative, but is anticipated on the client based on input. A history of previous player inputs is kept, so when a new update comes in from the server, the player object reapplies inputs from the frames since the "anticipation tick" in order to compute a new responsive anticipated position.
  - This is set to smooth from previous prediction to new prediction over 0.1 seconds, but there is a slider that can change that between 0 seconds (snap to new position) and a full second of interpolation. 0.1 feels good and smooth and responsive though.

Launching a second client, for values that have smoothing enabled (the third and fourth network variable above and the player), the second client will interpolate from current position to new position when it receives a server update. This results in pretty nice smooth movement of the player even though the transform has interpolation disabled.

### Issue Number(s)
[MTT-8052](https://jira.unity3d.com/browse/MTT-8052)

### Contribution checklist
 - [ ] Tests have been added for the project and/or any internal package
 - [ ] Release notes have been added to the [project changelog](../CHANGELOG.md) file
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] JIRA ticket ID is in the PR title or at least one commit message
 - [x] Include the ticket ID number within the body message of the PR to create a hyperlink

### CODE REVIEW

File: Experimental/Anticipation Sample/Assets/DefaultNetworkPrefabs.asset
Comment (fernando-cortez) [inline Experimental/Anticipation Sample/Assets/DefaultNetworkPrefabs.asset:1]: Since this sample doesn't spawn players, this can be removed, right? It references the BootstrapPlayer from Netcode for GameObjects package, which seems unused.
At the same time, we could disable the toggle in the Netcode for GameObjects project settings so that this file doesn't get auto-created.

File: Experimental/Anticipation Sample/Assets/Materials/Player.mat
Comment (fernando-cortez) [inline Experimental/Anticipation Sample/Assets/Materials/Player.mat:1]: On quick inspection, this material is unused as well, and could be removed.

File: Experimental/Anticipation Sample/README.md
Comment (fernando-cortez) [inline Experimental/Anticipation Sample/README.md:4]: This points to 1.7.1 still, however this upcoming version is unreleased.
Should we just point to the branch for now?
We could add a Jira ticket to point to 1.9.x when that is out.

File: Experimental/Anticipation Sample/Assets/Scripts/AnticipationSample.cs
Comment (fernando-cortez) [inline Experimental/Anticipation Sample/Assets/Scripts/AnticipationSample.cs:262]: Super small nitpick, but maybe we keep this slider near the variables section in the GUI?
Initially I thought this value had to do with the transform's smoothing, and not the sliders.

**Review by fernando-cortez — COMMENTED** [review]
Really like this!

A few bugs I saw while playtesting:
* Dragging the top bar (your local anticipated value) for sliders C and D can cause a desync between the local and server's value on another client. So on client A the sliders will look okay, but on client B, the sliders might be mismatched.
* We should be seeing a client's anticipated value for ValueE slider, right? Client's only see the server's authoritative value for E, but the value for E (in editor) looks to keep growing, and goes outside the bounds of 0-10.

There's a few other minor things I've commented on, but these seem like they should be fixed, or clarified.

**Comment by ShadauxCat** [comment]
@fernando-cortez 

> Dragging the top bar (your local anticipated value) for sliders C and D can cause a desync between the local and server's value on another client. So on client A the sliders will look okay, but on client B, the sliders might be mismatched.

Interesting. I hadn't tried that. I'll look into that.

> We should be seeing a client's anticipated value for ValueE slider, right? Client's only see the server's authoritative value for E, but the value for E (in editor) looks to keep growing, and goes outside the bounds of 0-10.

I left a comment about this in the code. The actual value for the variable does not stay within the bounds of 0-10, even on the server, but is modulo by 10 for the display. I did it that way because having it actually jump from 10 back to 0 suddenly created a difficult situation on the client: either the client would have to always smooth, in which case the server would teleport back to 0 but the client would interpolate back to 0, or the client would have to use some heuristic to determine when not to smooth. I wasn't able to come up with a heuristic for that that didn't cause some odd behavior near the boundary. So I decided to make the modulo happen in the display of the value rather than happening in the actual value.

File: Experimental/Anticipation Sample/Assets/Scripts/AnticipationSample.cs
Comment (ShadauxCat) [reply Experimental/Anticipation Sample/Assets/Scripts/AnticipationSample.cs:262]: Fair.

**Comment by ShadauxCat** [comment]
I actually realized my issue with variable E could be trivially solved if I could see the previous authoritative value in the callback (then I could teleport if the authoritative value went down and smooth otherwise) so I'm looking into adding that to the reanticipate callback, at which point I can change value E to be constrained between 0 and 10 properly. Thanks for that comment, it prompted me to think about that situation again and made me uncover a good change to better support user needs.

**Comment by ShadauxCat** [comment]
Nevermind, that actually didn't help at all.

So there's two ways I can handle the rollover:

- I can check if the new authority value is less than the previous authority value and teleport... this doesn't work because it delays the jump to after I've already predicted it, which causes a hitch.
- I can check if the new predicted value is less than the previous predicted value... this doesn't work because the "previous predicted value" is the current display value, so if it's in the middle of smoothing, this is the value at the low end of the interpolation, so when I teleport it back to the beginning, it interrupts the interpolation and results in it teleporting back before it reaches the end and then hitching a little at the beginning.

What this actually needs is for the smooth algorithm to handle the teleport, which is a bit more complicated, but not impossible, so I'm looking into that now.

File: Experimental/Anticipation Sample/Assets/Materials/Player.mat
Comment (ShadauxCat) [reply Experimental/Anticipation Sample/Assets/Materials/Player.mat:1]: It is used by the Player in-scene-placed object.

Comment (ShadauxCat) [reply Experimental/Anticipation Sample/Assets/Materials/Player.mat:1]: Apparently I forgot to check in the update to the scene where I actually used that material.

File: Experimental/Anticipation Sample/Assets/DefaultNetworkPrefabs.asset
Comment (ShadauxCat) [reply Experimental/Anticipation Sample/Assets/DefaultNetworkPrefabs.asset]: done

File: Experimental/Anticipation Sample/Assets/Scripts/AnticipationSample.cs
Comment (ShadauxCat) [reply Experimental/Anticipation Sample/Assets/Scripts/AnticipationSample.cs]: done

**Comment by ShadauxCat** [comment]
Fixed the two bugs you reported (and a third one I found while testing it) in the feature branch in Netcode for GameObjects.

**Comment by Elfi0Kuhndorf** [comment]
I think it is a great sample to showcase client anticipation. :)
I think it would be good to be able to change jitter and latency throughout runtime, to showcase how that influences the behavior.
image

**Comment by Elfi0Kuhndorf** [comment]
Also it is small optimization, but I noticed that the Anticipated Network Transform transmits the scale values of the game objects, I think it could be unticked as the scale value never changes or? Should save a tiny bit of bandwidth.

**Comment by ShadauxCat** [comment]
> I think it is a great sample to showcase client anticipation. :) I think it would be good to be able to change jitter and latency throughout runtime, to showcase how that influences the behavior. image

I wanted to do this, honestly! Unfortunately the simulator doesn't seem to allow changing the simulation parameters while the transport is connected. You have to disconnect, change the parameters, and reconnect.

Though, maybe I can add an "apply" button at runtime that will do exactly that...

**Comment by ShadauxCat** [comment]
> Also it is small optimization, but I noticed that the Anticipated Network Transform transmits the scale values of the game objects, I think it could be unticked as the scale value never changes or? Should save a tiny bit of bandwidth.

Thanks for mentioning this... the moment I tried this, I found that when I uncheck axes to limit replication, AnticipatedNetworkTransform doesn't initialize correctly. I'll be figuring out why that is tomorrow.

File: Experimental/Anticipation Sample/README.md
Comment (ShadauxCat) [reply Experimental/Anticipation Sample/README.md]: done

**Comment by ShadauxCat** [comment]
Fixed that initialization issue. I think all the feedback has been applied now.

**Comment by fernando-cortez** [comment]
Hello @ShadauxCat, I retested things just now and I'm seeing the mismatch on slider C and D still, and also not seeing the client's anticipated value for E. Should this be the case?

**Comment by ShadauxCat** [comment]
@fernando-cortez I don't see that on my machine. I did have to make updates in the NGO branch to fix things, is it possible that your NGO branch might be out of date? I'm not sure how or when Unity handles fetching updates to packages specified with git urls.

**Comment by fernando-cortez** [comment]
@ShadauxCat Ah yes, Package Manager won't fetch the latest value automatically.
We have to make sure that packages-lock.json's hash value is modified and points to the latest.
I'll retest with it locally, but I'll let you modify that!

**Comment by fernando-cortez** [comment]
It's all working with the new hash!
I'll approve when that hash is updated! 😄

**Comment by ShadauxCat** [comment]
Hash updated! I want to make a small update to the readme too before this merges in. I added support for performing smoothing on the server side (because I realized it would be really hard to make a host client feel smooth without support for that) and I want to make sure it's mentioned.

**Review by fernando-cortez — DISMISSED** [review]
Good to go!

**Comment by Elfi0Kuhndorf** [comment]
image
I sometimes get this null reference or the one below when exiting playmode 
image

**Comment by ShadauxCat** [comment]
> image I sometimes get this null reference or the one below when exiting playmode image

@Elfi0Kuhndorf - fixed these in the NGO branch.

**Pull request merged**
